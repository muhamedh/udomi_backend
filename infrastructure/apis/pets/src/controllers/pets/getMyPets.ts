import middy from "@middy/core";
import {
  APIGatewayProxyEvent,
  APIGatewayProxyResult,
  Context,
} from "aws-lambda";

import { QueryCommand, QueryCommandOutput } from "@aws-sdk/client-dynamodb";
import { unmarshall } from "@aws-sdk/util-dynamodb";
import { docClient } from "../../clients/dynamodb";
import { getUserId } from "../../helpers/jwt";

export const getMyPets = middy().handler(
  async (
    event: APIGatewayProxyEvent,
    context: Context
  ): Promise<APIGatewayProxyResult> => {
    console.log("Proceeding with my pet fetch");

    const userId = getUserId(event?.headers?.Authorization);
    const command = new QueryCommand({
      TableName: process.env.USER_TABLE,
      KeyConditionExpression: "id = :id",
      ExpressionAttributeValues: {
        ":id": { S: userId },
      },
    });
    try {
      let data = await docClient.send(command);
      return {
        statusCode: 200,
        body: data.Items ? JSON.stringify(data.Items) : ([] as any),
      };
    } catch (e) {
      console.log(`Database fetch failed with message ${e}`);
      throw Error("Database fetch failed");
    }
  }
);
