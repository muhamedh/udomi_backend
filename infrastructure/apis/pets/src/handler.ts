import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { PutCommand, DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb";
import { v4 as uuidv4 } from "uuid";
import { APIGatewayProxyEvent } from "aws-lambda";
import jwt from "jsonwebtoken";

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);
const getUserId = (token: any): any => {
  try {
    const decoded = jwt.decode(token, { complete: true });
    if (decoded) {
      return decoded.payload.sub;
    } else {
      console.log("Not decoded");
      throw new Error("Invalid token");
    }
  } catch (error) {
    console.log(error);
    throw new Error("Invalid token");
  }
};
const addPet = async (userId: any, pet: any) => {
  console.log("Proceeding with pet insert");
  const command = new PutCommand({
    TableName: process.env.USER_TABLE,
    Item: {
      id: userId,
      pet_id: `PET_ID#${uuidv4()}`,
      data: pet ? JSON.stringify(pet) : {},
    },
  });
  try {
    await docClient.send(command);
  } catch (e) {
    console.log(`Database insert failed with message ${e}`);
    throw Error("Database insert failed");
  }
};
export const handler = async (event: APIGatewayProxyEvent, context: any) => {
  console.log("Received event:", JSON.stringify(event, null, 2));
  try {
    if (
      event?.httpMethod === "POST" &&
      event?.resource === "/pets" &&
      event?.headers?.Authorization
    ) {
      const userId = getUserId(event?.headers?.Authorization);
      await addPet(userId, event?.body);
    } else {
      return {
        isBase64Encoded: false,
        statusCode: 404,
        body: JSON.stringify({
          message: "Invalidna metoda.",
        }),
      };
    }

    return {
      isBase64Encoded: false,
      statusCode: 200,
      body: JSON.stringify({
        message: "Ljubimac uspješno dodan.",
      }),
    };
  } catch (error) {
    return {
      isBase64Encoded: false,
      statusCode: 500,
      body: JSON.stringify({
        message: "Greška na serveru.",
      }),
    };
  }
};
