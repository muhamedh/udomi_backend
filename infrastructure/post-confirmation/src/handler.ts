import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { PutCommand, DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb";

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

export const handler = async (event: any, context: any) => {
  //console.log("Received event:", JSON.stringify(event, null, 2));
  //console.log("Proceeding with database insert");
  const command = new PutCommand({
    TableName: process.env.USER_TABLE,
    Item: {
      id: event?.request?.userAttributes?.sub,
      pet_id: "PET_ID#",
      email: event?.request?.userAttributes?.email,
    },
  });
  
  try {
    const response = await docClient.send(command);
  } catch (e) {
    console.log(`Database insert failed with message ${e}`);
  }
  return context.done(null, event);
};
