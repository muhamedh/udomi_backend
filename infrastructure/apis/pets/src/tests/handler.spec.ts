import { getMyPetsEvent } from "./mocks/events/getMyPets";
import { handler } from "../handler";
import {
  nockDynamoDBGetMyPetsCall,
  nockDynamoDBGetMyPetsErrorCall,
} from "./mocks/nockHelper";
import nock from "nock";

//   nock.recorder.rec();

describe("GET /pets/my", () => {
  jest.setTimeout(15000);
  const event = getMyPetsEvent;
  const context = {};
  afterAll(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    nock.cleanAll();
  });

  it("Should return 200 status code and user's pets", async () => {
    nockDynamoDBGetMyPetsCall();

    const response = await handler(getMyPetsEvent, context as any, jest.fn());

    expect(response).toMatchSnapshot();
  });
  it("Should return 5XX status code when database call fails", async () => {
    nockDynamoDBGetMyPetsErrorCall();
    try {
      const response = await handler(getMyPetsEvent, context as any, jest.fn());
      expect(true).toBe(false);
    } catch (e) {
      expect(true).toBe(true);
    }
  });
});
