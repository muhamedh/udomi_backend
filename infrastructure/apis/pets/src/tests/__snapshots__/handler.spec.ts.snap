// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /pets/my Should return 200 status code and user's pets 1`] = `
{
  "body": "[{"email":{"S":"dummy@email.com"},"id":{"S":"dummy-user-id"},"data":{"S":"{\\"phoneNumber\\":\\"2312312\\"}"},"pet_id":{"S":"PET_ID#"}},{"id":{"S":"dummy-user-id"},"data":{"S":"{\\"petName\\":\\"Dummy Pet Name\\",\\"location\\":\\"Dummy Location\\",\\"vaccinatedStatus\\":\\"VACCINATED\\",\\"chippedStatus\\":\\"CHIPPED\\",\\"shortDescription\\":\\"Dummy Short Description\\"}"},"pet_id":{"S":"PET_ID#8312d226-7c90-4ec4-a4df-8def0ece1bcb"}}]",
  "headers": {
    "Access-Control-Allow-Headers": "Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token",
    "Access-Control-Allow-Methods": "OPTIONS,GET,POST",
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;
